internal/util/util.go

```diff
@@ -19,19 +19,96 @@ package util
 import (
 	"crypto/sha1"
 	"fmt"
+	"sort"
 
-	"helm.sh/helm/v3/pkg/chartutil"
+	goyaml "gopkg.in/yaml.v2"
 	"helm.sh/helm/v3/pkg/release"
+	"sigs.k8s.io/yaml"
 )
 
 // ValuesChecksum calculates and returns the SHA1 checksum for the
 // given chartutil.Values.
-func ValuesChecksum(values chartutil.Values) string {
-	var s string
-	if len(values) != 0 {
-		s, _ = values.YAML()
+func ValuesChecksum(values map[string]interface{}) string {
+	newValues := copyValues(values)
+	var (
+		s   []byte
+		err error
+	)
+
+	if len(newValues) != 0 {
+		// cleanUpInterfaceMap
+		for i, value := range newValues {
+			newValues[i] = cleanUpMapValue(value)
+		}
+		msValues := yaml.JSONObjectToYAMLObject(newValues)
+		// Sort
+		SortMapSlice(msValues)
+		// Marshal
+		s, err = goyaml.Marshal(msValues)
+		if err != nil {
+			panic(err)
+		}
+	}
+	// Gethash
+	return fmt.Sprintf("%x", sha1.Sum(s))
+}
+
+func SortMapSlice(ms goyaml.MapSlice) {
+	sort.Slice(ms, func(i, j int) bool {
+		return fmt.Sprint(ms[i].Key) < fmt.Sprint(ms[j].Key)
+	})
+	for _, item := range ms {
+		if nestedMS, ok := item.Value.(goyaml.MapSlice); ok {
+			SortMapSlice(nestedMS)
+		} else if _, ok := item.Value.([]interface{}); ok {
+			for _, vItem := range item.Value.([]interface{}) {
+				if itemMS, ok := vItem.(goyaml.MapSlice); ok {
+					SortMapSlice(itemMS)
+				}
+			}
+		}
+	}
+}
+
+func cleanUpMapValue(v interface{}) interface{} {
+	switch v := v.(type) {
+	case []interface{}:
+		return cleanUpInterfaceArray(v)
+	case map[interface{}]interface{}:
+		return cleanUpInterfaceMap(v)
+	case string:
+		return v
+	default:
+		return fmt.Sprintf("%v", v)
```

Comment:
The result of this would be that everything becomes a string type. This may have other undesirable effects, e.g. the change of an integer type to a string (or vice-versa) causing an upgrade to not be triggered.